### PR TITLE
inline compiler

### PR DESCRIPTION
**Pull request by jerch**
Wouldn't it be nice to declare small wasm functions inline?

Something along this:
```typescript
const func = WasmFunc('int', [['int', 'a'], ['int', 'b']], {`
  return a + b;
`});

console.log(func(23, 42)); // --> should print 65
```

### CODE REVIEW

**Comment by jerch** [comment]
Slowly getting somewhere:

`src/emwasm` contains a cleaner rewrite of the wasm compiler script. The folder will be moved into its own package later on. Ultimately it should bring 3 primitives for adding inline wasm stuff:
- `EmWasmBytes` for just embedding a wasm ressource, returns the wasm bytes (sync) or a promise (async)
- `EmWasmBytes` for embedding a wasm module (precompiled), returns the wasm module (sync) or a promise (async)
- `EmWasmInstance` for returning a ready-to-go instance on access (again instance or promise)

Currently only `EmWasmInstance` is halfway done, still alot stuff missing on the cli/compiler side.

**Comment by jerch** [comment]
Restructured EmWasmBytes/EmWasmModule/EmWasmInstance to a single EmWasm call with a type setting instead, which simplified several aspects of the compiler script and shrinks the output code by a few kB.

**Comment by jerch** [comment]
I am not happy yet, how the code replacements work - currently it scans the input file for enclosing sentinel tokens `##EM_WASM##` and `##\EM_WASM##` in comments, and does a partial module execution to get proper content resolving and match the content from error stack position back to the definitions. While this works, it combines the worst of two worlds (macro like src code replacements + partial real execution) creating several bad coding restrictions.

Needs some investigations into these directions:
- would tagged template literals with a clever tag function make things easier to digest?
- real AST parsing, e.g. with acorn?

**Comment by jerch** [comment]
Things left to do:
- [ ] proper type inference from byte/module intermediates
- [ ] lazy emscripten sdk interactions (pull in first usage etc.)
- [ ] multi compiler backend support (at least clang directly from emscripten, zig? rust? AS?)

**Comment by jerch** [comment]
With the last commit the compiler script properly deals with exceptions from different wasm feature sets, eg. this is now possible:
```typescript
const wasmInstance = (() => {
  try {
    return EmWasm({ /* SIMD code */ });  // will throw on webkit browsers
  } catch (e) {
    return EmWasm({ /* scalar code */ });
  }
})();
wasmInstance.exports.some_func(...);
```
Still important to note here, that more complicated control flow around `EmWasm` calls should be avoided, since the compile script has to enter all conditional paths to see calls to `EmWasm`. For the case above this is guaranteed by the fact, that the compile script itself throws an `EmWasmReadExit` error. Even with a simple branching from some random predicate this wont work anymore.

Therefore higher level runtime branching as (eg. with `wasm-feature-detect`) must be done afterwards. For this only `OutputType.BYTES` can be used safely with `EmWasm`, and instantiation has to be done manually.

**Comment by jerch** [comment]
@Tyriar: Do you know of an easy way to overload type stuff from default namespace? I have the following issue:

`new WebAssembly.Module(...)` returns a `WebAssembly.Module`, which I want to extend in a way to carry forward the wasm definitions (used for imports/exports properties on the final instance). I already have dummy types for that, but cant get the type overload working.

As an interim solution I just provide an interface for `WebAssembly`, so manually attaching that gets the job done:
```typescript
interface WebAssemblyExt {
  Module: WebAssembly.Module & {
    new<T extends any>(bytes: (T extends IWasmDefinitionSyncBytes ? WasmBytes<T> : Uint8Array)):
      T extends IWasmDefinition ? WasmModule<T> : WebAssembly.Module;
  };
  ...
}

const someModule = new (WebAssembly as WebAssemblyExt).Module(someBytes);
// type --> WebAssembly.Module

const typedBytes: WasmBytes<XY> = ...;
const typedModule = new (WebAssembly as WebAssemblyExt).Module(typedBytes);
// type --> WasmModule<XY>
```

WebAssembly's base type is defined in `lib.dom.d.ts` as a namespace, but the common namespace extension idioms dont work, as `Module` is a statically typed `var` here https://github.com/microsoft/TypeScript/blob/bb913f83400c49727a3f666e5f8d07fea3b54c2a/lib/lib.dom.d.ts#L17808

Any idea how to get this done w'o re-interfacing the whole namespace?

**Comment by jerch** [comment]
@Tyriar Found a simpler way to solve the issue above by redeclaring the namespace, schematically:

```typescript
// definitions.ts
export declare namespace WebAssemblyExtended {
  ... // copied from lib.dom.ts + custom function overloads
}

// usage somewhere in the code
import { WebAssemblyExtended } from 'definitions';

(WebAssembly as typeof WebAssemblyExtended) // contains now overloaded type declarations
```
This is still somewhat ugly as it means to maintain a copy of a default namespace, but better than any other trickery I've found so far.

**Comment by jerch** [comment]
with the last 2 changes the wasm bootstrapping penalty got much better:
- `EmWasm` now returns a getter function, which removes the base64 decoding / module compilation / instantiation penalty from initial loading phase (not eval'ed before calling the getter)
- `EmWasm` memorizes decoded bytes and the wasm module (act as "singletons"), which reduces runtime for multiple getter calls later on. We dont memorize instances though for `OutputType.INSTANCE`, as this would mean to have only one global instance at runtime. If that behavior is desired, it can easily be shimmed by keeping the first instance around.

**Comment by jerch** [comment]
With the last changes the basic functionality is implemented and manually tested. Next steps from here on:
- [ ] rework package structure:
  - [ ] make it a cli tool package
  - [ ] rename to `emwasm` (embedded wasm) to `inwasm` (inline wasm) to avoid ambiguity with emscripten naming scheme
  - [ ] make cli tool to index.ts (remove old example code)
- [ ] type tests (with `tsd`?)
- [ ] better zig integration (auto bootstrap similar to emscripten?)
- [ ] compiler backend tests for:
  - [ ] emscripten-c/c++
  - [ ] clang-c/c++
  - [ ] zig
  - [ ] rust
- [ ] basic usage docs

**Comment by Tyriar** [comment]
> Found a simpler way to solve the issue above by redeclaring the namespace

This is similar to what I normally do when redeclaring `window`:

https://github.com/xtermjs/xterm.js/blob/a8405b86defb32743c3834a19741cbc117bbda5e/demo/client.ts#L36-L48

**Comment by jerch** [comment]
> This is similar to what I normally do when redeclaring `window`:

Hmm, gonna try an interfaced version again. For some reason I did not get it working properly with the ctor overloads, only the function overloads worked with that. Maybe I was holding it wrong? The interface variant is much easier/nicer to work with, that a full naemspace re-definition.

File: testproject/src/inline.wasm.ts [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+  },
+  code: `
+  #[no_mangle]
+  pub extern fn doubled(x: i32) -> i32 {
+    x * 2
+  }
+  `
+})();
+console.log(rust.exports.doubled(66));
+console.log(rust.exports.doubled(-333));
```
[the remark this answers is not in the document]
Comment (jerch) [reply testproject/src/inline.wasm.ts:233]: Well it currently needs the rust toolchain to be installed, and cargo in PATH. Need to get config file support done first, before I can mess with autoinstaller functionality. Currently Zig and Rust need to be preinstalled, emscripten kinda already supports autoinstalling, but the sdk path resolution needs better abstraction from config options.

[the remark this answers is not in the document]
Comment (jerch) [reply testproject/src/inline.wasm.ts:233]: Yes that should do to get started with. Zig on the other hand is less convenient to install. (You may wonder why I added it at all - well it has a great c compiler and support async/await itself, thus makes writing async code alot easier.)

[the remark this answers is not in the document]
Comment (Tyriar) [reply testproject/src/inline.wasm.ts:233]: I wouldn't want to take on keeping installation docs/scripts perpetually up to date for multiple languages is all, definitely feels out of scope of this.

[the remark this answers is not in the document]
Comment (jerch) [reply testproject/src/inline.wasm.ts:233]: Yes thats true - well it is a tradeoff between easy usage/convenience (main goal is to keep `npm install` working on random TS source packages, that may contain `InWasm`), and to give enough room for advanced compiler setups (never gonna work with simple `npm install`). Not sure yet, how far I will push the auto setup stuff, for rust at least it is quite tricky (due `rustup` mainly being an interactive script).

**Pull request merged**